### ISSUE TITLE
Inconsistent definition of a valid solution node

### ISSUE THREAD

**Issue by YanSong97** [zhentingqi/rStar#13]
Hi there,

Thank you for the excellent work on this project!

While reviewing the code, I noticed an inconsistency in how valid solution nodes are defined within the `Reasoning_MCTS_Node` class. Specifically, in the `is_valid_leaf_node()` function, a valid solution node includes types `SUBQUESTION`, `DIRECT_ANSWER`, or `OST_STEP`. However, in the `extract_solution_from_node()` function finding the best solution, the `OST_STEP` type is excluded.

Is there any reason for this exclusion?

Many thanks

**Comment by beishangeyu** [comment]
I have same question, and I've noticed that in the `is_terminal` function, the `OST_STEP` is excluded as well

**Comment by YanSong97** [comment]
> I have same question, and I've noticed that in the `is_terminal` function, the `OST_STEP` is excluded as well

Yeah, seems weird. FYI, we have re-implemented the reasoning part in this [PR](https://github.com/openreasoner/openr/pull/33). After rough testing, I believe adding `OST_STEP` to `is_terminal` condition can slightly improve performance.

**Comment by beishangeyu** [comment]
Many thanks! I will try in my experiment.

**Comment by Thunderbeee** [comment]
Thanks so much for your contribution and your PR! :D Toward your question why OST_STEP is excluded in the solution extractor, tldr, it is our design choice to let DIRECT_ANSWER (chain of thought) etc to be the final node to answer the question.

The is_valid_solution_node() function is likely used to determine if a node can be considered a valid solution during the search process. This might include nodes that are intermediate steps (OST_STEP) because they contribute to the reasoning process and might be necessary for the overall solution path.

The extract_solution_from_node() function, on the other hand, is used to extract the final solution from a node. In this context, OST_STEP nodes might not directly provide a final answer or solution, which is why they are excluded.
